### ISSUE TITLE
tiT blocking CPU1 and error stopping wifi (IDFGH-1293)

### ISSUE THREAD

**Issue by piltom** [espressif/esp-idf#3587]
# Environment

- Development Kit:      ESP32-DevKitC
- Kit version v4
- Module or chip used:  ESP32-WROOM-32D
- IDF version (run ``git describe --tags`` to find it): 
    v4.0-dev-667-gda13efc17
- Build System:      Make
- Compiler version: 1.22.0-80-g6c4433a
- Operating System:    Linux
- Power Supply:         USB

## Problem Description

When stopping soft ap, the following error appears on monitor:
`E (30352) tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014`

Later, when trying to connect in sta mode to another ap, the watchdog gets triggered because tiT (tcp/ip task) is hanging the cpu1:
```
E (42482) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:

E (42482) task_wdt:  - IDLE1 (CPU 1)

E (42482) task_wdt: Tasks currently running:

E (42482) task_wdt: CPU 0: IDLE0

E (42482) task_wdt: CPU 1: tiT
```

### Expected Behavior
Program description:
The program implements a state machine that loads the last pair of ssid/pass from nvs, if it can connect it sends a request and then turns off. If it can't connect, it stops the station mode, starts a soft ap with an http server for setting the ssid/pass. When that value pair is posted, it saves them to nvs, turns off soft ap, and goes back to the initial state (except nvs load, which is called only once on startup).

For connecting to an ap in sta mode, i implemented a component, which works on its own (tested). It receives ssid, pass, max number of tries and a callback pointer. It tries to connect an amount of times and if it fails, it calls the callback and then calls esp_wifi_stop().

For creating the http server, I implemented 2 components, one for the server itself and one for wifi in ap mode. When a combination of new ssid and pass is entered, the server shuts itself down and again calls esp_wifi_stop() (and a callback to notify the state machine of the new config)
### Actual Behavior
-> Try to connect to the saved ssid and pass, failed (as expected)
-> start soft ap and set up web server.
-> After entering the configuration, post request is received and data is processed
-> The function for starting a new connection to an ap in sta mode is started
->The problems described happen

### Code to reproduce this issue
I would have to ask for permission for this. I've logged a lot of events in the debug logs, it contains a lot of information.
Every time the handler for the connection is called, I log the event and event id.
The component for connecting to the wifi uses a modified version of the simple connect example.

## Debug Logs
What's going on on the log:
Trying 2 times to connect to the saved AP. Logged a WIFI_EVENT, with id 3 (sta stop)
After failing this, it creates the server, i connect and enter the values, pass them through a post request, and the values for ssid/pass are right (checked through logging).
Here starts the failing:
I call again for esp_wifi_stop(). Two WIFI_EVENT are logged, 3 and 14 (sta stop and ap sta dissconected).
Then i get this E log:
tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014

Then another WIFI_EVENT with id 12 (ap stop).
Then the same E log as before.
After this, the code for connecting to the ap starts again (which works if you give it a valid ssid/pass the first time).

```
I (28) boot: ESP-IDF v4.0-dev-667-gda13efc17 2nd stage bootloader
I (28) boot: compile time 16:10:31
I (28) boot: Enabling RNG early entropy source...
I (34) boot: SPI Speed      : 40MHz
I (38) boot: SPI Mode       : DIO
I (42) boot: SPI Flash Size : 4MB
I (46) boot: Partition Table:
I (50) boot: ## Label            Usage          Type ST Offset   Length
I (57) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (65) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (72) boot:  2 factory          factory app      00 00 00010000 00100000
I (80) boot: End of partition table
I (84) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x213f0 (136176) map
I (140) esp_image: segment 1: paddr=0x00031418 vaddr=0x3ffb0000 size=0x03074 ( 12404) load
I (146) esp_image: segment 2: paddr=0x00034494 vaddr=0x40080000 size=0x00400 (  1024) load
0x40080000: _WindowOverflow4 at /home/espdev/esp/esp-idf/components/freertos/xtensa_vectors.S:1779

I (148) esp_image: segment 3: paddr=0x0003489c vaddr=0x40080400 size=0x0b774 ( 46964) load
I (176) esp_image: segment 4: paddr=0x00040018 vaddr=0x400d0018 size=0x8db40 (580416) map
0x400d0018: _stext at ??:?

I (379) esp_image: segment 5: paddr=0x000cdb60 vaddr=0x4008bb74 size=0x04160 ( 16736) load
0x4008bb74: prvCopyDataToQueue at /home/espdev/esp/esp-idf/components/freertos/queue.c:2044

I (396) boot: Loaded app from partition at offset 0x10000
I (396) boot: Disabling RNG early entropy source...
I (396) cpu_start: Pro cpu up.
I (400) cpu_start: Application information:
I (405) cpu_start: Project name:     app-template
I (410) cpu_start: App version:      df62230-dirty
I (416) cpu_start: Compile time:     Jun  4 2019 16:10:37
I (422) cpu_start: ELF file SHA256:  f6085d5d9d98acfb...
I (428) cpu_start: ESP-IDF:          v4.0-dev-667-gda13efc17
I (434) cpu_start: Starting app cpu, entry point is 0x400811f0
0x400811f0: call_start_cpu1 at /home/espdev/esp/esp-idf/components/esp32/cpu_start.c:267

I (0) cpu_start: App cpu up.
I (445) heap_init: Initializing. RAM available for dynamic allocation:
I (451) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (457) heap_init: At 3FFB91F8 len 00026E08 (155 KiB): DRAM
I (464) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (470) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (476) heap_init: At 4008FCD4 len 0001032C (64 KiB): IRAM
I (483) cpu_start: Pro cpu start user code
I (165) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
Done
I (231) btnapp: set SSID:default set password:default
I (341) wifi station: connect to ap SSID:default password:default
I (341) wifi station: ESP_WIFI_MODE_STA
I (351) wifi: wifi driver task: 3ffb9e90, prio:23, stack:3584, core=0
I (351) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE
I (351) system_api: Base MAC address is not set, read default base MAC address from BLK0 of EFUSE
I (381) wifi: wifi firmware version: e11ccc4
I (381) wifi: config NVS flash: enabled
I (381) wifi: config nano formating: disabled
I (381) wifi: Init dynamic tx buffer num: 32
I (391) wifi: Init data frame dynamic rx buffer num: 32
I (391) wifi: Init management frame dynamic rx buffer num: 32
I (401) wifi: Init management short buffer num: 32
I (401) wifi: Init static rx buffer size: 1600
I (401) wifi: Init static rx buffer num: 10
I (411) wifi: Init dynamic rx buffer num: 32
I (521) phy: phy_version: 4100, 2a5dd04, Jan 23 2019, 21:00:07, 0, 0
I (531) wifi: mode : sta (30:ae:a4:d3:3a:b4)
I (531) wifi station: eventbase WIFI_EVENT, event id 2
I (531) wifi station: wifi_init_sta finished.
I (531) wifi station: connect to ap SSID:default password:default
I (2581) wifi station: eventbase WIFI_EVENT, event id 5
I (2591) wifi station: retry to connect to the AP
I (2591) wifi station: connect to the AP fail
I (4641) wifi station: eventbase WIFI_EVENT, event id 5
I (4641) wifi station: retry to connect to the AP
I (4651) wifi: flush txq
I (4651) wifi: stop sw txq
I (4651) wifi: lmac stop hw txq
I (4651) wifi station: connect to the AP fail
I (4651) wifi softAP: ESP_WIFI_MODE_AP
I (4651) wifi station: eventbase WIFI_EVENT, event id 3
W (4671) event: handler already registered, overwriting
W (4671) event: handler already registered, overwriting
I (4711) wifi: mode : softAP (30:ae:a4:d3:3a:b5)
I (4711) wifi: Total power save buffer number: 16
I (4711) wifi: Init max length of beacon: 752/752
I (4711) wifi: Init max length of beacon: 752/752
I (4721) wifi softAP: wifi_init_softap finished. SSID:Config Server password:defpass
I (4721) wifi station: eventbase WIFI_EVENT, event id 11
I (4741) config_server: Starting server on port: '80'
I (4821) config_server: Registering URI handlers
I (19281) wifi: new:<1,1>, old:<1,0>, ap:<1,1>, sta:<255,255>, prof:1
I (19281) wifi: station: 00:28:f8:79:f5:95 join, AID=1, bgn, 40U
I (19331) wifi station: eventbase WIFI_EVENT, event id 13
I (19331) wifi softAP: station 00:28:f8:79:f5:95 join, AID=1
I (19361) tcpip_adapter: softAP assign IP to station,IP is: 192.168.4.2
I (30611) config_server: Found header => Host: 192.168.4.1
I (30611) config_server: Request headers lost
W (30841) httpd_uri: httpd_uri: URI '/favicon.ico' not found
W (30841) httpd_txrx: httpd_resp_send_err: 404 Not Found - This URI does not exist
I (37061) config_server: echo post handler, 28
I (37061) config_server: =========== RECEIVED DATA ==========
I (37061) config_server: ssid=maskaw&pass=dOmNJDfu
I (37071) config_server: ====================================
I (37071) config_server: rec SSID:maskaw rec password:dOmNJDfu
I (37081) wifi: station: 00:28:f8:79:f5:95 leave, AID = 1, bss_flags is 134243
I (37091) wifi: new:<1,0>, old:<1,1>, ap:<1,1>, sta:<255,255>, prof:1
I (37091) wifi station: eventbase WIFI_EVENT, event id 3
I (37101) wifi station: eventbase WIFI_EVENT, event id 14
E (37101) tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014
I (37111) wifi station: eventbase WIFI_EVENT, event id 12
E (37121) tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014
I (37131) config_server: Stopping webserver
I (37531) wifi: flush txq
I (37531) wifi: stop sw txq
I (37531) wifi: lmac stop hw txq
I (37821) wifi station: connect to ap SSID:maskaw password:dOmNJDfu
I (37821) wifi station: ESP_WIFI_MODE_STA
W (37821) event: handler already registered, overwriting
W (37821) event: handler already registered, overwriting
W (37841) event: handler already registered, overwriting
I (37891) wifi: mode : sta (30:ae:a4:d3:3a:b4)
I (37891) wifi station: eventbase WIFI_EVENT, event id 2
I (37891) wifi station: wifi_init_sta finished.
I (37891) wifi station: eventbase WIFI_EVENT, event id 2
I (37901) wifi station: connect to ap SSID:maskaw password:dOmNJDfu
I (38501) wifi: new:<6,0>, old:<1,0>, ap:<255,255>, sta:<6,0>, prof:1
I (39491) wifi: state: init -> auth (b0)
I (39501) wifi: state: auth -> assoc (0)
I (39511) wifi: state: assoc -> run (10)
I (43531) wifi: state: run -> init (2a0)
I (43531) wifi: new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1
I (43531) wifi: new:<6,0>, old:<6,0>, ap:<255,255>, sta:<6,0>, prof:1
E (49501) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (49501) task_wdt:  - IDLE1 (CPU 1)
E (49501) task_wdt: Tasks currently running:
E (49501) task_wdt: CPU 0: IDLE0
E (49501) task_wdt: CPU 1: tiT

```

[sdkconfig.zip](https://github.com/espressif/esp-idf/files/3256086/sdkconfig.zip)

**Comment by piltom** [comment]
I'm setting up a JTAG debugger at the moment, will update.

**Comment by piltom** [comment]
Suspending the program during the watchdog trigger suggests that the tcp/ip task might be stuck in [igmp_tmr](http://www.nongnu.org/lwip/2_0_x/igmp_8c.html#a4418a22d37098f05d0c2fcfe288d0ca1).

The error could be related to [this](http://lwip.100.n7.nabble.com/TCP-tmr-endless-loop-td4091.html).

I will continue to debug, however i feel like this kind of error should be prevented somehow internally in esp-idf. If I find any more data, i'll update.

**Comment by liuzfesp** [comment]
Hi @piltom, it's our bug, we will fix it recently.

**Comment by liuzfesp** [comment]
HI @piltom I think it should be a bug, we will fix it and update you.

**Comment by piltom** [comment]
> 
> 
> HI @piltom I think it should be a bug, we will fix it and update you.

Is there any news?

**Comment by liuzfesp** [comment]
For the error log:

> tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014

Hopefully the fix for this error message will be merged into master. However, the error log is NOT related to the task watchdog, I guess the task watchdog is caused by other reason. Currently I just can't replicate the task watchdog issue described above. Is the task watchdog easy  reproduced on your side, or it's very difficult to be reproduced?

**Comment by liuzfesp** [comment]
@piltom, could you help to do following test:

- Enable task watchdog panic via menuconfig: make menuconfig => Component Config => Common ESP-related => Enable "Invoke panic handler on Task Watchdog timeout" 
- Build the image and reproduce the task watchdog issue
- Get backtrace via "xtensa-esp32-elf-addr2line -piaf -e build/*.elf backtrace_log" to get the backtrace that causes taskwatchdog.

**Comment by piltom** [comment]
> 
> 
> For the error log:
> 
> > tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014
> 
> Hopefully the fix for this error message will be merged into master. However, the error log is NOT related to the task watchdog, I guess the task watchdog is caused by other reason. Currently I just can't replicate the task watchdog issue described above. Is the task watchdog easy reproduced on your side, or it's very difficult to be reproduced?

It gets triggered all the time. tiT got stuck in an infinite loop, probably due to the wifi error. Unfortunately it's been a month since i posted the issue and i don't have the code anymore.

**Comment by torresnt** [comment]
Hello,
I am having the same issue:
********************************************************************************
I (96357) wifi: station: b8:ee:65:bc:99:b7 leave, AID = 1, bss_flags is 134243
I (96357) wifi: new:<1,0>, old:<1,1>, ap:<1,1>, sta:<0,0>, prof:1
main: (152012) event got: 0x00000000
sys: (152012) Processing event, type: 0, id: 17
sys: (152012) SYSTEM_EVENT
sys: (152012) AP_STADISCONNECTED
server: (151956) Stopping Server
I (96497) Device HTTP: Server Stop
wifi: (161768) Disconnecting AP 
server: AP disconnected
sys: (161876) Connecting to: xxxx with password: yyyy
wifi: (161876) Setting WiFi configuration xxxx
wifi: (161876) wifi sta initialized
E (96527) tcpip_adapter: handle_ap_stop 134 esp_wifi_internal_reg_rxcb ret=0x3014
I (96537) wifi: flush txq
I (96537) wifi: stop sw txq
I (96537) wifi: lmac stop hw txq
I (96617) wifi: mode : sta (30:ae:a4:8f:cc:b4)
*****************************************************************************************
Any updates on it? It happens when my device, acting as a server (AP), is connected to a station (Web browser) in APSTA mode and I issue an esp_wifi_stop to restore the connection and connect my device in STA mode to an existing AP or router.

**Comment by liuzfesp** [comment]
Hi @piltom @torresnt this issue has been fixed:

- v3.3 - 9a1b11c8
- master - b7d6aa74
- v3.2 - c119a1a0
- v3.1 - 896d597a

Strongly recommend you to upgrade your branch to latest stable IDF releases.

**Comment by Alvin1Zhang** [comment]
@piltom Thanks for reporting, feel free to reopen if the issue still happens. Thanks.

**Comment by Amey112** [comment]
Hey, I am a beginner to ESP32 and I am using PlatformIO for development, I am getting the exact same problem using:
PACKAGES:
 - framework-arduinoespressif32 3.10006.210326 (1.0.6)
 - tool-esptoolpy 1.30100.210531 (3.1.0)
 - toolchain-xtensa32 2.50200.97 (5.2.0)
Do you know of any workarounds?
